src/components/legend/draw.js

```diff
@@ -480,26 +480,76 @@ function setupTraceToggle(g, gd) {
 
 function handleClick(g, gd, numClicks) {
     if(gd._dragged || gd._editing) return;
+
     var hiddenSlices = gd._fullLayout.hiddenlabels ?
         gd._fullLayout.hiddenlabels.slice() :
         [];
 
-    var legendItem = g.data()[0][0],
-        fullData = gd._fullData,
-        trace = legendItem.trace,
-        legendgroup = trace.legendgroup,
-        traceIndicesInGroup = [],
-        tracei,
-        newVisible;
+    var legendItem = g.data()[0][0];
+    var fullData = gd._fullData;
+    var fullTrace = legendItem.trace;
+    var legendgroup = fullTrace.legendgroup;
+
+    var i, j, carr, key, keys, val;
+    var attrUpdate = {};
+    var attrIndices = [];
+    var carrs = [];
+    var carrIdx = [];
+
+    function insertUpdate(traceIndex, key, value) {
+        var attrIndex = attrIndices.indexOf(traceIndex);
+        var valueArray = attrUpdate[key];
+        if(!valueArray) {
+            valueArray = attrUpdate[key] = [];
+        }
+
+        if(attrIndices.indexOf(traceIndex) === -1) {
+            attrIndices.push(traceIndex);
+            attrIndex = attrIndices.length - 1;
+        }
+
+        valueArray[attrIndex] = value;
+
+        return attrIndex;
+    }
+
+    function setVisibility(fullTrace, visibility) {
+        var fullInput = fullTrace._fullInput;
+        if(Registry.hasTransform(fullInput, 'groupby')) {
+            var carr = carrs[fullInput.index];
+            if(!carr) {
+                var groupbyIndices = Registry.getTransformIndices(fullInput, 'groupby');
+                var lastGroupbyIndex = groupbyIndices[groupbyIndices.length - 1];
+                carr = Lib.keyedContainer(fullInput, 'transforms[' + lastGroupbyIndex + '].styles', 'target', 'value.visible');
+                carrs[fullInput.index] = carr;
+            }
+
+            // If not specified, assume visible:
+            var curState = carr.get(fullTrace._group) || true;
```

Comment:
Why does this happen?

Reply:
this `===` that `|| true` fallback.

Reply:
Good catch. This isn't a great conditional. If it's really `fullData`, it should *always* be present and what it means to be. I think I'll just remove this.

Reply:
Haha couldn't remove it. Modified it to be a bit more targeted/explicit. The issue is that the container doesn't always have `visible` defined. "Oh, so just use `fullTrace.visible`." That didn't work either. Broke some tests. Aimed for some middle ground that passes all the tests.